### ISSUE TITLE
3.16 - Bail out from beacon script if already processed or not acceptable screen size

### ISSUE THREAD

**Issue by Miraeld** [wp-media/wp-rocket#6432]
## Context
As a user, I want WP Rocket to perform an accurate analysis of LCP/ATF images so that I get the best optimization possible.This issue has the goal to develop the orange part of this diagram:
⚠ The diagram mentions that, for not cached pages, the check about existing data should be bypassed. After discussion with Piotr, we don't really care so we can simplify the flow and check in all cases.
<img width="1000" alt="Untitled" src="https://github.com/wp-media/wp-rocket/assets/14993267/16585f77-6365-4ef5-822d-8637ae32a305">

**Must be done after #6431  so that it can be tested**

## Scope a solution 
- To implement this, we need to modify the beacon script from #6431 and add a check if we already have images related to the URL within the database. If yes or the status is failed, stop the script process.
- And also add a screen-size check within the script:
    - if one of the dimension is below the one of Pixel 4 (used in RUCSS as mobile: 393 x 830), bail out.
    - if one of the dimension is above the one of 1080p (1920x1080), bail out. 

For the first check, we could add an additional AJAX call that would trigger a method inside the AJAX controller (callback on the subscriber) that would check if there are any records for the current URL (using the query class).
- Add tests

## Acceptance Criteria
Bail-out if data is already available:
- Make sure there is a log in the beacon script in case it bails out becasue data is already available.
- Clear cache and ATF/LCP data. Make sure cache feature and LCP/ATF feature are enabled.
- Visit a page from desktop.
- Check in the DB that the LCP/ATF data is available for this page.
- Visit the page once again from desktop. Check in the dev tool console that the script about bailing out for data already there is printed out.
- Visit the page from mobile. Check that there is no log about bailing out for data laready there.
- Check in the DB that the mobile LCP/ATF data has been added.

Bail-out if not acceptable screen-size: (Use BrowserStack or Rocket-E2E)
- Make sure there is a log in the script if it bails out because of screen size.
- Clear cache and LCP/ATF data.
- Visit a page with a screen-size 392 x 1000. The script should bail out with the right log.
- Visit a page with a screen-size 1000 x 829. The script should bail out with the right log. 
- Visit a page with a screen-size 1000 x 1081. The script should bail out with the right log. 
- Visit a page with a screen-size 1921 x 1080. The script should bail out with the right log. 
- Visit a page with a screen-size 1920 x 1080. The script should complete and data must be available in the DB.
- Clear LCP/ATF data.
- Visit a page with a screen-size 393 x 830. The script should complete and data must be available in the DB.
- Clear LCP/ATF data.
- Visit a page with a screen-size 1000 x 1000. The script should complete and data must be available in the DB.

Check the screen-size filters:
- Repeat the above test about screen-size after applying different upper and lower limits with a filter:
    - Just set the upper limit: repeat the test (adapting the emulated screen-size).
    - Just set the lower limit: repeat the test (adapting the emulated screen-size).
- Validate that the filter is protected against unexpected inputs (null, empty, text, etc...)

##  Effort estimation:
Effort [S]

**Comment by MathieuLamiot** [comment]
Some work is needed here on AC, to check with @Mai-Saad's [suggestion](https://www.notion.so/Beacon-Auto-optimize-LCP-image-Auto-exclude-images-above-the-fold-from-LazyLoad-56226bccec874b45aed4a57f08690029?d=1d69da22a5a848e98097b94e905b0ba9&pvs=4#5f6f2345124f4814a83db5f40d4e458a): 

> beacon will check if the screen size is acceptable one to prevent generating data for weird screen sizes, bailout if not

    ◦ Mobile size will be defined as ‘≤RUCSS mobile threshold’
    ◦ Desktop size will be defined as ‘≥RUCSS desktop threshold’
    ◦ Both values will be filterable

We'll need an AC for the case where the results are already in the DB.

**Comment by MathieuLamiot** [comment]
We should have AC / task about measuring the script execution time, especially on pages with a lot of images to ensure we remain efficient and optimized.

**Comment by MathieuLamiot** [comment]
I am adding this task to the service team sprint as disucssed during sprint planning, so that @mostafa-hisham can share his feedbacks on what needs to be improved from the prototype script as well: multiple for loops, some images are not handled, etc.

**Comment by mostafa-hisham** [comment]
@MathieuLamiot 
I added comments on the Beacon [PR](https://github.com/wp-media/nodejs-treeshaker/pull/579) to explain the improvements we can make to the beacon script 
https://github.com/wp-media/nodejs-treeshaker/pull/579/files

**Comment by MathieuLamiot** [comment]
Thanks @mostafa-hisham. I see there are points to be discussed with the Product team ; can you raise this point to them directly on Slack and share the discussion here so that the plugin team can pick it up as well?

**Comment by mostafa-hisham** [comment]
@MathieuLamiot Done and I added a replay on [my comment](https://github.com/wp-media/nodejs-treeshaker/pull/579/files#r1490604687).

> If Images are not on the viewport, they won't be worth optimizing as a critical asset, and won't bring any improvement for LCP.

[slack convo](https://wp-media.slack.com/archives/C015FT0RBKQ/p1707985409253389)

**Comment by MathieuLamiot** [comment]
@piotrbak According to the graph, the check about data being already here or not should be by-passed in case we are not on a cached page. I am not sure this can be done easily, and I am wondering why this is done here?
If we are serving a non-cached page, this check would have been done already at PHP level to know if we should inject or not the beacon. There could be a race condition where the results get in the DB between page generation and the script running, but is it a big issue? 
Is there a specific reason for always triggering the beacon on not-cached versions?

**Comment by hanna-meda** [comment]
~@piotrbak, This has been implemented up until the "Check the screen-size filters:" part. So this part is not covered by the current ticket:~

> Check the screen-size filters:

> Repeat the above test about screen-size after applying different upper and lower limits with a filter:
> Just set the upper limit: repeat the test (adapting the emulated screen-size).
> Just set the lower limit: repeat the test (adapting the emulated screen-size).
> Validate that the filter is protected against unexpected inputs (null, empty, text, etc...)

~Should this be done as part of 3.16? Either way, we need a separate GH to cover for it.~

fyi @MathieuLamiot & @Mai-Saad 

[Update] Disregard the above. In the end, filters have been added within the current ticket. They are documented in [here](https://www.notion.so/wpmedia/Beacon-Auto-optimize-LCP-image-Auto-exclude-images-above-the-fold-from-LazyLoad-66f9411b46b9414082ac985b5cb83fb1?pvs=4#6901f32f2ef34e78ab57ab674975d8ee). Thanks, @Miraeld.

**Comment by MathieuLamiot** [comment]
Follow-up on this @piotrbak: https://github.com/wp-media/wp-rocket/issues/6432#issuecomment-2020309717

Currently, the check to the DB if there is already or not LCP/ATF images, or if there is a failed row is done all the time, not only on cached version. either the documentation must be adapted, either a dedicated follow-up GH issue must be added. Thanks

**Comment by piotrbak** [comment]
@MathieuLamiot About this one, it was done to prevent checking the same thing twice. On the not-cached version we'll:
1. Check in PHP if there's entry related to this URL
2. Check using the beacon the same thing

I thought that checking if we're on the cached page would be cheaper in terms of resource usage than doing PHP/SQL processing. At the end, if you think this is not relevant, we could update the documentation. WDYT?

**Comment by MathieuLamiot** [comment]
Checking first if we're on a cached page would probably be a little bit more efficient, especially for websites with a lot of visits.
But the feature is working without ; so I don't know if we prefer to consider this for later enhancement, or have it in the initial release...

**Comment by piotrbak** [comment]
` especially for websites with a lot of visits`
Yes, that was my main concern here. We had an option to clear the cache after injecting the data (which was rejected) or go this way. I think we could go with enhancement here, or at least check the effort during the grooming session. @DahmaniAdame

**Comment by MathieuLamiot** [comment]
Ok, we'd need a dedicated GH issue then I think. Thanks 🙏

**Comment by MathieuLamiot** [comment]
On-going discussion about the screen size limitations: https://wp-media.slack.com/archives/CUT7FLHF1/p1713457206386399?thread_ts=1712752333.384509&cid=CUT7FLHF1

**Comment by piotrbak** [comment]
@Miraeld For the visibility:
We should bail-out when:
1. (mobile) We have a screen bigger than RUCSS mobile size (393 x 830) (but still WP Rocket generates the mobile cache for it)
2. (desktop) We have a screen smaller than RUCSS desktop size (1920x1080)

@DahmaniAdame @benorfaz This is the AC from the epic. As Adame mentioned in the thread, we could also bail-out when detecting very big screensize. What would be its dimensions? Do we want to add it initially in 3.16 or as enhncement?

**Comment by DahmaniAdame** [comment]
> Do we want to add it initially in 3.16 or as enhncement?

As enhancement to avoid any delays. 

The only requirement on release is for the limits to be controlled by filters, which I trust is already the case.

**Comment by MathieuLamiot** [comment]
Bringing back to In Progress @Miraeld

**Comment by MathieuLamiot** [comment]
@Miraeld I am having an issue with this PR. It bails out when browsing on a 1920x1080 screen. I think this is because what we are measuring in the script is not the screen size but he display size, which only accounts for the size available to display the page, excluding the URL bar, the favorite bar, etc. So it can vary a bit for every user.

@piotrbak @DahmaniAdame Shouldn't we have some margin to account for this? On my screen, I something like 1900 x 870.

Compared to [this](https://github.com/wp-media/wp-rocket/issues/6432#issuecomment-2065384783):

Maybe bail-out when:

- (mobile) We have a screen bigger than RUCSS mobile size (500 x 1000) (but still WP Rocket generates the mobile cache for it)
- (desktop) We have a screen smaller than RUCSS desktop size (1600x700)

**Comment by DahmaniAdame** [comment]
🤔 isn't the condition for it to be smaller than 1920x1080? The condition should be fine for anything smaller than this resolution, regardless of what eats the screen's realestate.

> (desktop) We have a screen smaller than RUCSS desktop size (1600x700)

That won't prevent the script from running on screens bigger 1920x1080, which would have more chances and give worst mismatch on ATF/LCP than having a smaller resolution by changing the windows size.

**Comment by MathieuLamiot** [comment]
> isn't the condition for it to be smaller than 1920x1080? That won't prevent the script from running on screens bigger 1920x1080, which would have more chances and give worst mismatch on ATF/LCP than having a smaller resolution by changing the windows size.

Well that's not the latest specification we got [here](https://github.com/wp-media/wp-rocket/issues/6432#issuecomment-2065384783). but it could be changed to _bailing out if bigger_, solving the issue too.
We just need a clear specification :D

**Comment by DahmaniAdame** [comment]
I think there is a misunderstanding indeed 😅

@piotrbak shouldn't this be bigger than 1920x1080? 

> (desktop) We have a screen _**smaller than**_ RUCSS desktop size (1920x1080)

Because it we bail while it's smaller, we are excluding the vast majority of desktop views 😅

**Comment by piotrbak** [comment]
@DahmaniAdame @MathieuLamiot This was the condition since very beginning to bail out when the sceen **is smaller** than what we have for RUCSS.

@DahmaniAdame what do you think should be the minimal processed size here?
I'm fine with proposal of Mathieu, at least for the beginning until we have some feedback.

**shouldn't this be bigger than 1920x1080?**
If we bail out for the bigger ones, we'll have tablets being processed which is something we wanted to avoid.

**Comment by DahmaniAdame** [comment]
Let's go with @MathieuLamiot's proposal and iterate as we get feedback on mismatches.

**Comment by piotrbak** [comment]
@MathieuLamiot Thanks for your feedback and apologies for the confusion here. We'll go with what you've proposed:

We **should bail-out** when:
- (mobile) We have a **display area bigger** than 393x830 (but still WP Rocket generates the mobile cache for it)
- (desktop) We have a **display area smaller** than 1600x700

**Comment by Miraeld** [comment]
Here is the new PR to modify the thresholds values: 
https://github.com/wp-media/wp-rocket/pull/6593
